src/main/java/com/google/api/generator/engine/ast/CastExpr.java

```diff
@@ -52,11 +52,15 @@ public CastExpr build() {
       // Only numeric types can be casted around.
       if (castType.isPrimitiveType() && exprType.isPrimitiveType()) {
         // Check that the types are either both booleans or both not booleans.
-        // TODO(xiaozhenliu): This is one of the places where we need to check the equality of boxed
-        // and primitive types.
         Preconditions.checkState(
             !(castType.equals(TypeNode.BOOLEAN) ^ exprType.equals(TypeNode.BOOLEAN)),
```

Comment:
Could you also please check if `boolean x = (Boolean) true` and the reverse are valid?

Reply:
Yes, verified in local files and unit tests, cast expression below are valid.
```
        boolean flag = (Boolean) true;
        Boolean flag1 = (boolean) true;
```
but since similar test case is covered by `validCastExpr_castBoxedTypeToPrimitive` and `validCastExpr_castPrimitiveToBoxedType`, we won't add extra test for this. Thanks

Reply:
The test case won't catch this, because the original check looks specifically at the primitive type. Could we please add a test case? 

With the `equals()` change, the logic above won't need to be changed either.

Reply:
I'm a little bit confused here, the test cases would be like below if I'm understanding correctly which are already covered:
```
  @Test
  public void validCastExpr_castPrimitiveToBoxedType() {
    // [Constructing] `(Boolean) true` where `true` is a primitive value that is `boolean` type.
    PrimitiveValue booleanValue = PrimitiveValue.builder().setValue("true").setType(TypeNode.BOOLEAN).build();
    ValueExpr valueExpr = ValueExpr.withValue(booleanValue);
    CastExpr.builder().setType(TypeNode.BOOLEAN_OBJECT).setExpr(valueExpr).build();
  }

  @Test
  public void validCastExpr_castBoxedTypeToPrimitive() {
     // [Constructing] `(boolean) flag` where `flag` is a object value that is `Boolean` type.
    Variable variable = Variable.builder().setName("flag").setType(TypeNode.BOOLEAN_OBJECT).build();
    VariableExpr variableExpr = VariableExpr.builder().setVariable(variable).build();
    CastExpr.builder().setType(TypeNode.BOOLEAN).setExpr(variableExpr).build();
  }
```
are we talking about the whole assignment expression? Thanks

Reply:
a draft PR for reference: https://github.com/googleapis/gapic-generator-java/pull/140/files#diff-59533acd4d8a07afff86d7de2cd8606dR106

Reply:
done.